synthtool/languages/python.py

```diff
@@ -52,3 +66,101 @@ def fix_pb2_grpc_headers(*, proto_root: str = "**/*_pb2_grpc.py") -> None:
         PB2_GRPC_HEADER,
         fr"{LICENSE}\n\n\g<1>\n\n\g<2>",  # add line breaks to avoid stacking replacements
     )
+
+
+def _get_help(filename: str) -> str:
+    return shell.run(["python", filename, "--help"]).stdout
+
+
+def _get_noxfile_metadata(sample_dir: Path) -> dict:
+    metadata: Dict[str, Any] = {}
+    with open(sample_dir / "noxfile.py.yml") as f:
+        metadata = yaml.load(f, Loader=yaml.SafeLoader) or {}
+
+    # set defaults
+    metadata["DEFAULT_VERSIONS"] = SAMPLES_VERSIONS
+    if "ignored_versions" not in metadata:
+        metadata["ignored_versions"] = IGNORED_VERSIONS
+    if "install_library_from_source" not in metadata:
+        metadata["install_library_from_source"] = False
+
+    return metadata
+
+
+def _get_sample_readme_metadata(sample_dir: Path) -> dict:
+    sample_readme = sample_dir / "README.rst.in"
+
+    sample_metadata = {}
+    if sample_readme.exists():
+        requirements = str(Path(sample_dir / "requirements.txt").resolve())
+        log.debug(
+            f"Installing requirements at {requirements} to generate {sample_readme}"
+        )
+        shell.run([sys.executable, "-m", "pip", "install", "-r", requirements])
+
+        with open(sample_readme) as f:
+            sample_metadata = yaml.load(f, Loader=yaml.SafeLoader)
+        for sample in sample_metadata["samples"]:
+            # add absolute path to metadata so `python foo.py --help` succeeds
+            sample["abs_path"] = Path(sample_dir / (sample["file"])).resolve()
+
+    return sample_metadata
+
+
+def py_samples(*, root: PathOrStr = None, skip_readmes: bool = False) -> None:
+    """
+    Find all samples projects and render templates.
+    Samples projects always have a 'requirements.txt' file and may also have
+    README.rst.in
+
+    Args:
+        root (Union[Path, str]): The samples directory root.
+        skip_readmes (bool): If true, do not generate readmes.
+    """
+    in_client_library = Path("samples").exists() and Path("setup.py").exists()
+    if root is None:
+        if in_client_library:
+            root = "samples"
+        else:
+            root = "."
+
+    excludes = ["noxfile.py.yml"]
+    if skip_readmes:
+        excludes += [
+            str(p.relative_to(SAMPLES_TEMPLATE_PATH))
+            for p in SAMPLES_TEMPLATE_PATH.glob("*.rst")
+        ]
+    t = templates.TemplateGroup(SAMPLES_TEMPLATE_PATH, excludes=excludes)
```

Comment:
@SurferJeffAtGoogle Is it alright to use `templates` directly or will it break something related to the synth metadata tracking?